runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClient.java

```diff
@@ -88,34 +92,48 @@ public void close() {
     closeAndTerminateOutstandingRequests(new IllegalStateException("Runner closed connection"));
   }
 
+  public String getWorkerId() {
+    return workerId;
+  }
+
   /** Closes this client and terminates any outstanding requests exceptionally. */
   private void closeAndTerminateOutstandingRequests(Throwable cause) {
     if (isClosed.getAndSet(true)) {
       return;
     }
 
-    // Make a copy of the map to make the view of the outstanding requests consistent.
-    Map<String, CompletableFuture<BeamFnApi.InstructionResponse>> outstandingRequestsCopy =
-        new ConcurrentHashMap<>(outstandingRequests);
-    outstandingRequests.clear();
+    try {
+      // Make a copy of the map to make the view of the outstanding requests consistent.
+      Map<String, CompletableFuture<BeamFnApi.InstructionResponse>> outstandingRequestsCopy =
+          new ConcurrentHashMap<>(outstandingRequests);
+      outstandingRequests.clear();
 
-    if (outstandingRequestsCopy.isEmpty()) {
-      requestReceiver.onCompleted();
-      return;
-    }
-    requestReceiver.onError(
-        new StatusRuntimeException(Status.CANCELLED.withDescription(cause.getMessage())));
-
-    LOG.error(
-        "{} closed, clearing outstanding requests {}",
-        FnApiControlClient.class.getSimpleName(),
-        outstandingRequestsCopy);
-    for (CompletableFuture<BeamFnApi.InstructionResponse> outstandingRequest :
-        outstandingRequestsCopy.values()) {
-      outstandingRequest.completeExceptionally(cause);
+      if (outstandingRequestsCopy.isEmpty()) {
+        requestReceiver.onCompleted();
+        return;
+      }
+      requestReceiver.onError(
+          new StatusRuntimeException(Status.CANCELLED.withDescription(cause.getMessage())));
+
+      LOG.error(
+          "{} closed, clearing outstanding requests {}",
+          FnApiControlClient.class.getSimpleName(),
+          outstandingRequestsCopy);
+      for (CompletableFuture<BeamFnApi.InstructionResponse> outstandingRequest :
+          outstandingRequestsCopy.values()) {
+        outstandingRequest.completeExceptionally(cause);
+      }
+    } finally {
+      if (onCloseListener != null) {
+        onCloseListener.accept(this);
+      }
     }
   }
 
+  public void onClose(Consumer<FnApiControlClient> onCloseListener) {
```

Comment:
What happens if multiple layers try to attach a listener? I think this should either call both of them, or fail loudly.

Reply:
Changed to call all the registered listeners on close.